src/Heap.sol

```diff
@@ -36,209 +40,202 @@ library BasicHeap {
+            }
+
+            if (rightChildIndex < _size) {
+                Account memory rightChild = _heap.accounts[rightChildIndex];
+                if (rightChild.value > childToSwap.value) {
+                    childToSwap = rightChild;
+                    childIndex = rightChildIndex;
+                }
+            }
+
+            if (childToSwap.value > valueToShift) {
+                setAccount(_heap, childToSwap, _index);
+                _index = childIndex;
+                childIndex = (childIndex << 1) + 1;
+            } else {
+                break;
+            }
         }
-        setAccount(_heap, _rank, initialAccount);
+
+        setAccount(_heap, _accountToShift, _index);
     }
 
-    /// @notice Moves an account down the heap until its value is greater than the ones of its children.
-    /// @dev This functions restores the invariant about the order of the values stored when the account at `_rank` is the only one with value smaller than what it should be.
-    /// @param _heap The heap to modify.
-    /// @param _rank The index of the account to move.
-    function shiftDown(Heap storage _heap, uint256 _rank) private {
-        uint256 size = getSize(_heap);
-        Account memory initialAccount = getAccount(_heap, _rank);
-        uint256 initialValue = initialAccount.value;
-        Account memory childAccount;
-        uint256 childRank = _rank * 2;
-        // At this point, childRank (resp. childRank+1) is the rank of the left (resp. right) child.
-
-        while (childRank <= size) {
-            // Compute the rank of the child with largest value.
-            if (
-                childRank < size &&
-                getAccount(_heap, childRank + 1).value > getAccount(_heap, childRank).value
-            ) childRank++;
-
-            childAccount = getAccount(_heap, childRank);
-
-            if (childAccount.value > initialValue) {
-                setAccount(_heap, _rank, childAccount);
-                _rank = childRank;
-                childRank *= 2;
-            } else break;
+    /// @notice Checks if an account with the given address `_id` exists in the `_heap`.
+    /// @dev The parameters `_index`, `_accountsLength`, and `_id` must be coherent.
+    /// @param _heap The heap to search in.
+    /// @param _index The index to search for the account in.
+    /// @param _accountsLength The length of the `_heap` accounts array.
+    /// @param _id The address of the account to search for.
+    /// @return True if the account exists in the `_heap`, false otherwise.
+    function containsAccount(Heap storage _heap, uint256 _index, uint256 _accountsLength, address _id)
+        private
+        view
+        returns (bool)
+    {
+        if (_index != 0) {
+            return true;
+        } else if (_accountsLength != 0) {
+            return _heap.accounts[0].id == _id;
+        } else {
+            return false;
```

Comment:
```suggestion
        return _index != 0 || _heap.accounts[0].id == _id;
```
This is because if the length is 0 then the first account should be empty. We could even have something like
```suggestion
        return _heap.accounts[_index].id == _id;
```
but I'm guessing that it's less efficient

Reply:
We can't implement the last suggestion because it would revert if the length is zero.

Regarding the first suggestion, I initially tried that approach, but I was too afraid that the compiler might not always short-circuit boolean expressions (if the compiler evaluates both expressions in `_index != 0 || _heap.accounts[0].id == _id`, it can cause the function to revert). That's why I opted for the current implementation. I believe it makes the code more understandable and doesn't increase the gas cost.